firmware/components/buses/buses.c

```diff
@@ -0,0 +1,318 @@
+        #ifdef CONFIG_PIN_PULL_I2C0_DATA
+            i2c0BusConfiguration.sda_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c0BusConfiguration.sda_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+        #ifdef CONFIG_PIN_PULL_I2C0_CLK
+            i2c0BusConfiguration.scl_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c0BusConfiguration.scl_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+
+        res = i2c_param_config(I2C_NUM_0, &i2c0BusConfiguration);
+        if (res != ESP_OK) return res;
+        res = i2c_driver_install(I2C_NUM_0, i2c0BusConfiguration.mode, I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE, 0);
+        if (res != ESP_OK) return res;
+        
+        i2c0_mux = xSemaphoreCreateMutex();
+        if (i2c0_mux == NULL) return ESP_ERR_NO_MEM;
+    #endif
+
+    #ifdef CONFIG_BUS_I2C1_ENABLE
+        i2c_config_t i2c1BusConfiguration = {0};
+        i2c1BusConfiguration.mode             = I2C_MODE_MASTER;
+        i2c1BusConfiguration.sda_io_num       = CONFIG_PIN_NUM_I2C1_DATA;
+        i2c1BusConfiguration.scl_io_num       = CONFIG_PIN_NUM_I2C1_CLK;
+        i2c1BusConfiguration.master.clk_speed = CONFIG_I2C1_MASTER_FREQ_HZ,
+        #ifdef CONFIG_PIN_PULL_I2C1_DATA
+            i2c1BusConfiguration.sda_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2c1BusConfiguration.sda_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+        #ifdef CONFIG_PIN_PULL_I2C1_CLK
+            i2c1BusConfiguration.scl_pullup_en = GPIO_PULLUP_ENABLE,
+        #else
+            i2cqBusConfiguration.scl_pullup_en = GPIO_PULLUP_DISABLE,
+        #endif
+
+        res = i2c_param_config(I2C_NUM_1, &i2c1BusConfiguration);
+        if (res != ESP_OK) return res;
+        res = i2c_driver_install(I2C_NUM_1, i2c1BusConfiguration.mode, I2C_MASTER_RX_BUF_DISABLE, I2C_MASTER_TX_BUF_DISABLE, 0);
+        if (res != ESP_OK) return res;
+        
+        i2c1_mux = xSemaphoreCreateMutex();
+        if (i2c1_mux == NULL) return ESP_ERR_NO_MEM;
+    #endif
+
+    return ESP_OK;
+}
+
+/* I2C helper functions */
+
+esp_err_t driver_i2c_read_bytes(int bus, uint8_t addr, uint8_t *value, size_t value_len) {
+    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    esp_err_t res = i2c_master_start(cmd);
+    if (res != ESP_OK) { i2c_cmd_link_delete(cmd); return res; }
+    res = i2c_master_write_byte(cmd, ( addr << 1 ) | READ_BIT, ACK_CHECK_EN);
+    if (res != ESP_OK) { i2c_cmd_link_delete(cmd); return res; }
+    if (value_len > 1) {
+        res = i2c_master_read(cmd, value, value_len-1, ACK_VAL);
+        if (res != ESP_OK) { i2c_cmd_link_delete(cmd); return res; }
+    }
+    res = i2c_master_read_byte(cmd, &value[value_len-1], NACK_VAL);
+    if (res != ESP_OK) { i2c_cmd_link_delete(cmd); return res; }
+    res = i2c_master_stop(cmd);
+    if (res != ESP_OK) { i2c_cmd_link_delete(cmd); return res; }   
+
+    xSemaphoreHandle mux = (bus == I2C_NUM_1) ? i2c1_mux : i2c0_mux;
```

Comment:
This semaphore prevents two threads from trying to execute I2C actions at the same time, thus avoiding communication errors (needed: according to the IDF documentation the I2C functions of the SDK are not thread save).

Note: we should either add checks for this mutex to the MicroPython I2C driver or extend modi2c.c to a usable state and disable the built in I2C support of MicroPython to avoid bus conflicts which are happening on the current firmware when the user tries to communicate with a sensor attached to grove/sao connectors on their badge from within Python.